### PR TITLE
Initial commit

### LINKED ISSUE #2103: http_request task
## Use Case
Bolt should ship with a task that makes it easy to make web requests.

For now, we don't have to focus on having this be callable from a task, just from plans.

## Describe the Solution You Would Like

params:
base_url - fully qualified URL scheme
path - path to append to URL. 
method - get, put, post, delete (default to get)
body - optional
headers - hash of header name:value
cacert - for trusting the remote end, maybe create an SSL obj? (nice to have)
cert/key - present for client auth (low priority)
json_endpoint- whether or not to send conten_ type headers and serialize/parse JSON from the bodies.

It would be a bonus if this task was usable as a remote task or a local task merging base_url, headers, and supported SSL options from `_target` into the params.

### PR DESCRIPTION
**Pull request by beechtom**
This adds a `http_request` task which makes HTTP requests.

Closes puppetlabs/bolt#2103

### CODE REVIEW

File: tasks/init.json
Comment (beechtom) [inline tasks/init.json]: Went with `url` over a combination of `base_url` and `path`, as it seems simpler to only have a single parameter that handles the URL and plans can use interpolation to accomplish what `base_url` and `path` would.

File: tasks/init.json
Comment (beechtom) [inline tasks/init.json]: Opted for `body_format` over `json_endpoint`, as it provides a little more flexibility in indicating if the body should be serialized/parsed as JSON while allowing for other common formats as well.

File: tasks/init.json
Comment (beechtom) [inline tasks/init.json]: Added this since it seems helpful for users who may want to log when Bolt is making requests.

File: tasks/init.rb
Comment (beechtom) [inline tasks/init.rb]: `body_format` will set a default `Content-Type` header, which can be overridden by setting the same header under `headers`. I'm not sure if it's more appropriate to check the `body_format` or `Content-Type` header when determining how to format the body...

File: tasks/init.rb
Comment (adreyer) [inline tasks/init.rb]: A common pain point in dealing with JSON APIs is hitting the wrong route and getting a non-json response. I think it would be better if there was a way to make that an error rather than a completely unexpected structure for the response.

File: tasks/init.rb
Comment (nicklewis) [inline tasks/init.rb]: The task should always return the response as long as one was returned and leave it up to the caller to decide which responses are acceptable.

**Review by nicklewis — CHANGES_REQUESTED** [review]
I think this has a bit too much magic when what we really need _first_ is a low-level implementation of the HTTP protocol on which we can build abstractions / conveniences.

For now it might be best to leave building the body and setting the content-type, as well as parsing the response, up to the user. We can build an opinionated JSON in+out task or option separately.

File: tasks/init.json
Comment (adreyer) [reply tasks/init.json]: The purpose of that spec was to support remote targets where the url is configured in inventory and the path is specified for each invocation.

Comment (nicklewis) [reply tasks/init.json]: But wouldn't that be under `_target` then?

Comment (nicklewis) [reply tasks/init.json]: Ah meaning you would only pass `path` as a task parameter and omit `baseurl`, opting instead to use what was specified in `_target`?

File: tasks/init.json
Comment (nicklewis) [inline tasks/init.json]:
```suggestion
      "type": "Optional[Hash[String, String]]"
```

This causes an extremely obtuse Puppet error, for the record... "Syntax error at end of input" with no context.

File: tasks/init.rb
Comment (nicklewis) [inline tasks/init.rb]: `slice` doesn't work on Ruby < 2.5. Since this can run on a remote target rather than with Bolt's Ruby, we should aim for better compatibility.

I also ran into an issue with `compact` in the `ruby_task_helper`. That method doesn't work on Ruby < 2.4.

**Review by nicklewis — CHANGES_REQUESTED** [review]
There's some wonkiness around content encoding. For instance, hitting `https://www.google.com` returns a response with content-type indicating the encoding is `ISO-8859-1` but the Ruby string is binary. Forcing the encoding to ISO-8859-1 and then encoding it to UTF-8 works, but otherwise Bolt will complain.

On the other hand, `https://www.yahoo.com` doesn't specify an encoding but the Ruby string is UTF-8.

We may also want to have an option to follow redirects, as it's a hassle when you just get back a 301 Moved.

File: tasks/init.json
Comment (beechtom) [reply tasks/init.json]: 🤦

**Comment by beechtom** [comment]
Updated. The task will now enforce encoding for the body based on the `Content-Type` header (if present) before encoding as UTF-8. Also added an option for following redirects (default to `true`) and an option for setting the max number of redirects (default to `20`, which seems to be a common maximum used in browsers).

I'll look at getting the task helper updated to not use `compact`.

File: CHANGELOG.md
Comment (lucywyman) [inline CHANGELOG.md]: Probably want this to be `0.1.0` for the initial release https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase

File: Gemfile
Comment (lucywyman) [inline Gemfile:23]: Add pdk

File: metadata.json
Comment (lucywyman) [inline metadata.json]: Same update, 0.1.0

File: CHANGELOG.md
Comment (beechtom) [reply CHANGELOG.md]: I have the initial release at `1.0.0` since I'm assuming once this is in it's immediately going to be shipped with Bolt and be ready for use in production: https://semver.org/#how-do-i-know-when-to-release-100 . If we don't immediately ship this and are still considering it 'in development', then I'm fine going down to a `0.x` release.

File: init.json
Comment (lucywyman) [inline init.json]: This seems to be here by accident?

Comment (beechtom) [reply init.json]: Oops, had a duplicate clone of the repo and accidentally pushed from it. Should be better now. @lucywyman

File: tasks/init.json
Comment (lucywyman) [inline tasks/init.json:2]: It'd be good to specify input_method stdin here, if only to exemplify best practices (and also because we don't use env vars)

File: tasks/init.rb
Comment (lucywyman) [inline tasks/init.rb:7]: If we expect this to be used in plans and not directly - or really even if we expect it to be used directly - I think it makes sense to wrap this in a plan that ensures these gems are available.

File: CHANGELOG.md
Comment (lucywyman) [reply CHANGELOG.md]: Right but _we_ aren't currently using it in production, and it seems better to make as few guarantees as possible.

Comment (beechtom) [reply CHANGELOG.md]: I don't really understand what you mean by we aren't using it in production...
Regardless, updated to 0.1.0.

**Review by lucywyman — APPROVED** [review]

**Comment by lucywyman** [comment]
@adreyer or @nicklewis Does either of you want to re-review this before merge and release?

**Pull request merged**
